sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java

```diff
@@ -0,0 +1,452 @@
+     * Make list request list messages response.
+     *
+     * @param client the client
+     * @param hl7v2Store the hl 7 v 2 store
+     * @param filter the filter
+     * @param pageToken the page token
+     * @return the list messages response
+     * @throws IOException the io exception
+     */
+    public static ListMessagesResponse makeListRequest(
+        HealthcareApiClient client,
+        String hl7v2Store,
+        @Nullable String filter,
+        @Nullable String pageToken)
+        throws IOException {
+      return client.makeHL7v2ListRequest(hl7v2Store, filter, pageToken);
+    }
+
+    @Override
+    public Iterator<List<String>> iterator() {
+      return new HL7v2MessageIDPagesIterator(this.client, this.hl7v2Store, this.filter);
+    }
+
+    /** The type Hl7v2 message id pages iterator. */
+    public static class HL7v2MessageIDPagesIterator implements Iterator<List<String>> {
+
+      private final String hl7v2Store;
+      private final String filter;
+      private HealthcareApiClient client;
+      private String pageToken;
+      private boolean isFirstRequest;
+
+      /**
+       * Instantiates a new Hl 7 v 2 message id pages iterator.
+       *
+       * @param client the client
+       * @param hl7v2Store the hl 7 v 2 store
+       * @param filter the filter
+       */
+      HL7v2MessageIDPagesIterator(
+          HealthcareApiClient client, String hl7v2Store, @Nullable String filter) {
+        this.client = client;
+        this.hl7v2Store = hl7v2Store;
+        this.filter = filter;
+        this.pageToken = null;
+        this.isFirstRequest = true;
+      }
+
+      @Override
+      public boolean hasNext() throws NoSuchElementException {
+        if (isFirstRequest) {
+          try {
+            ListMessagesResponse response = makeListRequest(client, hl7v2Store, filter, pageToken);
+            List<String> msgs = response.getMessages();
+            if (msgs == null) {
+              if (response.get("hl7V2Messages") != null) {
+                return ((ArrayList<ArrayMap<String, String>>) response.get("hl7V2Messages")).size()
+                    > 0;
+              }
+              return false;
+            } else {
+              return !msgs.isEmpty();
+            }
+          } catch (IOException e) {
+            throw new NoSuchElementException(
+                String.format(
+                    "Failed to list first page of HL7v2 messages from %s: %s",
+                    hl7v2Store, e.getMessage()));
+          }
+        }
+        return this.pageToken != null;
+      }
+
+      @Override
+      public List<String> next() throws NoSuchElementException {
+        try {
+          ListMessagesResponse response = makeListRequest(client, hl7v2Store, filter, pageToken);
+          this.isFirstRequest = false;
+          this.pageToken = response.getNextPageToken();
+          List<String> msgs = response.getMessages();
+          if (msgs == null && response.get("hl7V2Messages") != null) {
```

Comment:
Note, this `ArrayMap<String, String>` is basically the `Message` model.
If you use the alpha API to query an beta HL7v2 store, there will be a single key `name` with the message ID value.
If you use the alpha API to query an alpha HL7v2 store, there will be all of the keys in the `Message` model.

Reply:
This abstract modeling of the Message would work just fine.